OpenCascade links to unavailable libtbb.12.8.dylib

### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
HOMEBREW_VERSION: 4.0.14
ORIGIN: https://github.com/Homebrew/brew
HEAD: ac0663ae5d2d4c76bb62363592d77dbad49ac30c
Last commit: 7 days ago
Core tap JSON: 23 Apr 10:18 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.10 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: quad-core 64-bit penryn
Clang: 13.0.0 build 1300
Git: 2.40.0 => /usr/local/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 11.7.6-x86_64
CLT: 13.2.0.0.1.1638488800
Xcode: N/A
```

```
Your system is ready to brew.
```

### Verification

- [X] My "`brew doctor` output" says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [X] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

I want to compile my software [LibrePCB](https://github.com/LibrePCB/LibrePCB) which now depends on the OpenCascade library. So beside installing the whole C++/CMake/Qt toolchain, I install OpenCascade with `brew install opencascade`. The library is then successfully found by CMake. Then I start compiling with `make`.

### What happened (include all command output)?

The build fails with the following error:

    make[2]: *** No rule to make target `/usr/local/opt/tbb/lib/libtbb.12.8.dylib', needed by `apps/librepcb-cli/librepcb-cli.app/Contents/MacOS/librepcb-cli'.  Stop.

The complete build log is available [here](https://dev.azure.com/LibrePCB/LibrePCB/_build/results?buildId=2989&view=logs&j=d71c4c56-574b-5ba9-9268-feb3539ccbff&t=6fe8d1f1-9628-53f7-acc1-66d9cd4a27c8).

### What did you expect to happen?

`libtbb.dylib` was automatically installed as a dependency of the `opencascade` package, so the build should succeed.

But it seems the `tbb` package installed `libtbb.12.9.dylib` instead of `libtbb.12.8.dylib` so it is no surprise the build fails.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew install git cmake qt5 opencascade
brew link --force qt5
git clone -b 3d-package-models --recursive https://github.com/LibrePCB/LibrePCB.git
cd LibrePCB && mkdir build && cd build
cmake ..
make -j8
```

With the following workaround, building LibrePCB succeeds:

```
  ln -s "/usr/local/opt/tbb/lib/libtbb.12.9.dylib" "/usr/local/opt/tbb/lib/libtbb.12.8.dylib"
  ln -s "/usr/local/opt/tbb/lib/libtbbmalloc.2.9.dylib" "/usr/local/opt/tbb/lib/libtbbmalloc.2.8.dylib"
```

But I'd expect `brew` to install the correct version of dependencies. Or how can I ensure the installed `opencascade` version is compatible with its dependency `tbb`? Note that I am not a macOS user and thus not experienced with the whole ecosystem (including Homebrew).

Please let me know if you need any additional information.

> How about a small CMake project we can build with CMake that reproduces the error you saw?

I thought a command without requiring any files is simpler, but OK here's a real test case:

Create file `CMakeLists.txt`:

```
cmake_minimum_required(VERSION 3.5)
project(test LANGUAGES CXX)
find_package(OpenCASCADE REQUIRED)
add_executable(test main.cpp)
target_include_directories(test SYSTEM PRIVATE "${OpenCASCADE_INCLUDE_DIR}")
target_link_libraries(test PRIVATE TKernel)
```

Create file `main.cpp`:

```
#include <Quantity_Color.hxx>
#include <Standard_Version.hxx>
#include <iostream>
int main(void) {
  Quantity_Color c;  // test linking
  std::cout << "OCCT Version: " << OCC_VERSION_COMPLETE << std::endl;
  return 0;
}
```

Build & run the application:

```
cmake .
make
./test
```

> Also, OpenCascade has now been rebuilt. You may need to do brew update && brew reinstall opencascade to pick up the new build.

Perfect, thank you! Here's the proof that both the test case and the update work like a charm :smiley: 

```
macos@macoss-iMac-Pro Documents % cmake .
-- The CXX compiler identification is AppleClang 13.0.0.13000029
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (1.0s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/macos/Documents
macos@macoss-iMac-Pro Documents % make
[ 50%] Building CXX object CMakeFiles/test.dir/main.cpp.o
make[2]: *** No rule to make target `/usr/local/opt/tbb/lib/libtbb.12.8.dylib', needed by `test'.  Stop.
make[1]: *** [CMakeFiles/test.dir/all] Error 2
make: *** [all] Error 2
macos@macoss-iMac-Pro Documents % brew update && brew reinstall opencascade
Updated 1 tap (homebrew/core).
==> New Formulae
ares                                                                                                                    millet
==> Fetching opencascade
==> Downloading https://ghcr.io/v2/homebrew/core/opencascade/manifests/7.7.0-1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/opencascade/blobs/sha256:75ab9f445fd48c02bf5b0f4318337bcbd3a51d9d87ee0dcf1d4094a5c9efdc14
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:75ab9f445fd48c02bf5b0f4318337bcbd3a51d9d87ee0dcf1d4094a5c9efdc14?se=2023-04-24T09%3A30%3A00Z&sig=tZJKu0J%2Bt%2BWmUgOmYEmbzWqsc8o0ftoQzB40o8mGDYM%3D&sp=r&s
######################################################################## 100.0%
==> Reinstalling opencascade 
==> Pouring opencascade--7.7.0.big_sur.bottle.1.tar.gz
🍺  /usr/local/Cellar/opencascade/7.7.0: 9,603 files, 171.9MB
==> Running `brew cleanup opencascade`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
macos@macoss-iMac-Pro Documents % make                                     
-- Configuring done (0.1s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/macos/Documents
[ 50%] Linking CXX executable test
[100%] Built target test
macos@macoss-iMac-Pro Documents % ./test
OCCT Version: 7.7.0
```

--- a/Formula/opencascade.rb
+++ b/Formula/opencascade.rb
@@ -30,7 +30,7 @@ class Opencascade < Formula
     sha256 cellar: :any_skip_relocation, x86_64_linux:   "83f337ba6677432b4380ea30745a7b3eb5d343c24cf1fcb82c7387ee04b8d357"
   end
 
-  depends_on "cmake" => :build
+  depends_on "cmake" => [:build, :test]
   depends_on "doxygen" => :build
   depends_on "rapidjson" => :build
   depends_on "fontconfig"
@@ -89,5 +89,33 @@ class Opencascade < Formula
 
     # Discard the first line ("DRAW is running in batch mode"), and check that the second line is "1"
     assert_equal "1", output.split(/\n/, 2)[1].chomp
+
+    # Make sure hardcoded library name references in our CMake config files are valid.
+    # https://github.com/Homebrew/homebrew-core/issues/129111
+    # https://dev.opencascade.org/content/cmake-files-macos-link-non-existent-libtbb128dylib
+    (testpath/"CMakeLists.txt").write <<~CMAKE
+      cmake_minimum_required(VERSION 3.5)
+      project(test LANGUAGES CXX)
+      find_package(OpenCASCADE REQUIRED)
+      add_executable(test main.cpp)
+      target_include_directories(test SYSTEM PRIVATE "${OpenCASCADE_INCLUDE_DIR}")
+      target_link_libraries(test PRIVATE TKernel)
+    CMAKE
+
+    (testpath/"main.cpp").write <<~CPP
+      #include <Quantity_Color.hxx>
+      #include <Standard_Version.hxx>
+      #include <iostream>
+      int main() {
+        Quantity_Color c;
+        std::cout << "OCCT Version: " << OCC_VERSION_COMPLETE << std::endl;
+        return 0;
+      }
+    CPP
+
+    system "cmake", "-S", ".", "-B", "build"
+    system "cmake", "--build", "build"
+    ENV.append_path "LD_LIBRARY_PATH", lib if OS.linux?
+    assert_equal "OCCT Version: #{version}", shell_output("./build/test").chomp
   end
 end

